### ISSUE TITLE
Insufficient escaping of backslashes when generating LIKE queries

### ISSUE THREAD

**Issue by dsyalex** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1150]
### Steps to reproduce

Try to use StartsWith with a backslash, like in this example:
```c#
using System.Reflection;
using System.Collections.Generic;
using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
using Microsoft.EntityFrameworkCore.Query;
using Microsoft.EntityFrameworkCore;

using System;
using System.Linq;
namespace EFGetStarted
{
    // I take no credit for IQueryableExtensions, it comes from Nikolay Kostov and the Stack Overflow community at https://stackoverflow.com/a/44180537/1306662
    public static class IQueryableExtensions
    {
        public static string ToSql<TEntity>(this IQueryable<TEntity> query) where TEntity : class
        {
            var enumerator = query.Provider.Execute<IEnumerable<TEntity>>(query.Expression).GetEnumerator();
            var relationalCommandCache = enumerator.Private("_relationalCommandCache");
            var selectExpression = relationalCommandCache.Private<SelectExpression>("_selectExpression");
            var factory = relationalCommandCache.Private<IQuerySqlGeneratorFactory>("_querySqlGeneratorFactory");

            var sqlGenerator = factory.Create();
            var command = sqlGenerator.GetCommand(selectExpression);

            string sql = command.CommandText;
            return sql;
        }

        private static object Private(this object obj, string privateField) => obj?.GetType().GetField(privateField, BindingFlags.Instance | BindingFlags.NonPublic)?.GetValue(obj);
        private static T Private<T>(this object obj, string privateField) => (T)obj?.GetType().GetField(privateField, BindingFlags.Instance | BindingFlags.NonPublic)?.GetValue(obj);
    }

    public class FactionContext : DbContext
    {
        public DbSet<Faction> Factions { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder options)
            => options.UseMySql("server=localhost;userid=api;pwd=password;port=3306;database=api;sslmode=none");
    }

    public class Faction
    {
        public int Id { get; set; }
        public string Tag { get; set; }
    }

    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            using (var db = new FactionContext())
            {
                // Create
                Console.WriteLine("Inserting a new faction");
                db.Add(new Faction { Tag = "L\\-" });
                db.SaveChanges();

                // Read
                Console.WriteLine("Querying for a faction");
                var faction = db.Factions
                    .Where(f => f.Tag.StartsWith("L"))
                    .First();

                Console.WriteLine(faction.Tag);

                // Read
                Console.WriteLine("Querying for a faction");
                var query = db.Factions
                    .Where(f => f.Tag.StartsWith("L\\"));
                Console.WriteLine(query.ToSql());
                faction = query.First();

                Console.WriteLine(faction.Tag);
            }
        }
    }
}
```

Result:
```
Hello World!
Inserting a new faction
Querying for a faction
L\-
Querying for a faction
SELECT `f`.`Id`, `f`.`Tag`
FROM `Factions` AS `f`
WHERE `f`.`Tag` IS NOT NULL AND ((`f`.`Tag` LIKE CONCAT('L\\', '%')) AND (LEFT(`f`.`Tag`, CHAR_LENGTH(CONVERT('L\\' USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT('L\\' USING utf8mb4) COLLATE utf8mb4_bin))
Unhandled exception. System.InvalidOperationException: Sequence contains no elements
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.First[TSource](IQueryable`1 source)
   at EFGetStarted.Program.Main(String[] args) in /path/to/Program.cs:line 71
```
Instead of the exception what we should actually expect is `L\-` again - the match we tried to do was simply more specific.

### The issue
The MySQL documentation has this nasty line at https://dev.mysql.com/doc/refman/8.0/en/string-comparison-functions.html
>  Because MySQL uses C escape syntax in strings (for example, \\n to represent a newline character), you must double any \\ that you use in LIKE strings. For example, to search for \\n, specify it as \\\\n. To search for \\, specify it as \\\\\\\\; this is because the backslashes are stripped once by the parser and again when the pattern match is made, leaving a single backslash to be matched against.

The query above simplifies to something like this:
```
MariaDB [api]> SELECT * FROM Factions WHERE Tag LIKE 'L\\%' AND LEFT(Tag, CHAR_LENGTH('L\\')) = 'L\\';
Empty set (0.00 sec)
```

Whereas it should be more like this:
```
MariaDB [api]> SELECT * FROM Factions WHERE Tag LIKE 'L\\\\%' AND LEFT(Tag, CHAR_LENGTH('L\\')) = 'L\\';
+----+------+
| Id | Tag  |
+----+------+
|  1 | L\-  |
```

Note the extra LEFT check. This should not have extra escaping and therefore the caller cannot work around this bug by simply performing its own extra escaping.

### Further technical details

Tested with a few different versions of things:
MySQL version: 10.0.34-MariaDB-0ubuntu0.16.04.1 and 10.1.44-MariaDB-0ubuntu0.18.04.1
Operating system: Ubuntu 18.04 Bionic
Pomelo.EntityFrameworkCore.MySql version: 1.1.2, 3.1.2
Microsoft.AspNetCore.App version: 2.2.5, 3.1.7

**Comment by lauxjpn** [comment]
_(I only tested this with Pomelo `3.1.2`, since version `1.1.2` is way out of support as this point in time. Also, while unrelated to your issue here, MariaDB 10.0 and 10.1 are not officially [supported](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/#supported-dbms-and-versions) anymore either.)_

The following sample program works without issues using Pomelo `3.1.2` (tested with `8.0.20-mysql` and `10.1.43-mariadb`):

<details>
<summary>Sample program to demonstrate correct behavior</summary>

```c#
using System.Diagnostics;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace IssueConsoleTemplate
{
    public class Faction
    {
        public int Id { get; set; }
        public string Tag { get; set; }
    }
    
    public class Context : DbContext
    {
        public DbSet<Faction> Factions { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder
                .UseMySql(
                    "server=127.0.0.1;port=3308;user=root;password=;database=Issue1150",
                    b => b.ServerVersion("8.0.20-mariadb"))
                .UseLoggerFactory(
                    LoggerFactory.Create(
                        b => b
                            .AddConsole()
                            .AddFilter(level => level >= LogLevel.Information)))
                .EnableSensitiveDataLogging()
                .EnableDetailedErrors();
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<Faction>()
                .HasData(
                    new Faction { Id = 1, Tag = @"L\-M"});
        }
    }

    internal static class Program
    {
        private static void Main()
        {
            using var context = new Context();
            
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            var startsWith1 = context.Factions.FirstOrDefault(f => f.Tag.StartsWith(@"L"));
            var startsWith2 = context.Factions.FirstOrDefault(f => f.Tag.StartsWith(@"L\"));
            var equals = context.Factions.FirstOrDefault(f => f.Tag.Equals(@"L\-M"));
            var endsWith = context.Factions.FirstOrDefault(f => f.Tag.EndsWith(@"\-M"));
            var contains = context.Factions.FirstOrDefault(f => f.Tag.Contains(@"\-"));
            
            Debug.Assert(startsWith1 != null);
            Debug.Assert(startsWith2 != null);
            Debug.Assert(equals != null);
            Debug.Assert(endsWith != null);
            Debug.Assert(contains != null);
        }
    }
}
```

</details>

<details>
<summary>Generated SQL queries by sample program as logged by EF Core</summary>

```sql
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.3 initialized 'Context' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: ServerVersion 10.1.43 MariaDb SensitiveDataLoggingEnabled DetailedErrorsEnabled

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (3ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      CREATE DATABASE `Issue1150`;

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      CREATE TABLE `Factions` (
          `Id` int NOT NULL AUTO_INCREMENT,
          `Tag` longtext CHARACTER SET utf8mb4 NULL,
          CONSTRAINT `PK_Factions` PRIMARY KEY (`Id`)
      );

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (2ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      INSERT INTO `Factions` (`Id`, `Tag`)
      VALUES (1, 'L\\-M');

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (4ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` IS NOT NULL AND (`f`.`Tag` LIKE 'L%')
      LIMIT 1

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` IS NOT NULL AND (`f`.`Tag` LIKE 'L\\\\%')
      LIMIT 1

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` = 'L\\-M'
      LIMIT 1

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` IS NOT NULL AND (`f`.`Tag` LIKE '%\\\\-M')
      LIMIT 1

info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` LIKE '%\\\\-%'
      LIMIT 1
```

</details>

I can therefore **not reproduce** the issue described by you.

### Where to go from here

Please ensure, that you are actually using Pomelo `3.1.2`. Then, post the EF Core log output (configured in the same way as I have done above).

**Comment by dsyalex** [comment]
Interesting. Here's my csproj file:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.7">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="3.1.2" />
  </ItemGroup>

</Project>
```
Should I be looking in any other places?

**Comment by dsyalex** [comment]
I tried making a new directory with the following csproj:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.1.7">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="3.1.7" />
    <PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="3.1.2" />
  </ItemGroup>

</Project>
```
and your Program.cs, with only the connection string changed, and got this:
```
$ dotnet run
warn: Microsoft.EntityFrameworkCore.Model.Validation[10400]
      Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data, this mode should only be enabled during development.
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 3.1.7 initialized 'Context' using provider 'Pomelo.EntityFrameworkCore.MySql' with options: ServerVersion 10.1.43 MariaDb SensitiveDataLoggingEnabled DetailedErrorsEnabled 
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (62ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      DROP DATABASE `api`;
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      CREATE DATABASE `api`;
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (55ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      CREATE TABLE `Factions` (
          `Id` int NOT NULL AUTO_INCREMENT,
          `Tag` longtext CHARACTER SET utf8mb4 NULL,
          CONSTRAINT `PK_Factions` PRIMARY KEY (`Id`)
      );
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (5ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      INSERT INTO `Factions` (`Id`, `Tag`)
      VALUES (1, 'L\\-M');
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (2ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` IS NOT NULL AND ((`f`.`Tag` LIKE CONCAT('L', '%')) AND (LEFT(`f`.`Tag`, CHAR_LENGTH(CONVERT('L' USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT('L' USING utf8mb4) COLLATE utf8mb4_bin))
      LIMIT 1
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` IS NOT NULL AND ((`f`.`Tag` LIKE CONCAT('L\\', '%')) AND (LEFT(`f`.`Tag`, CHAR_LENGTH(CONVERT('L\\' USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT('L\\' USING utf8mb4) COLLATE utf8mb4_bin))
      LIMIT 1
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` = 'L\\-M'
      LIMIT 1
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE `f`.`Tag` IS NOT NULL AND (RIGHT(`f`.`Tag`, CHAR_LENGTH(CONVERT('\\-M' USING utf8mb4) COLLATE utf8mb4_bin)) = CONVERT('\\-M' USING utf8mb4) COLLATE utf8mb4_bin)
      LIMIT 1
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT `f`.`Id`, `f`.`Tag`
      FROM `Factions` AS `f`
      WHERE LOCATE(CONVERT('\\-' USING utf8mb4) COLLATE utf8mb4_bin, `f`.`Tag`) > 0
      LIMIT 1
Process terminated. Assertion failed.
   at IssueConsoleTemplate.Program.Main() in /path/to/PomeloFoundationEFMySQLBugReprod/Program.cs:line 57
```

So it fails at `Debug.Assert(startsWith2 != null);`?

**Comment by lauxjpn** [comment]
My apologies, as it turns out, I did not run the code with the `3.1.2` release, but with the current [nightly build](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#nightly-builds). The issue you described has already been fixed by #1057 and will be part of the upcoming `3.2.0` release.

Feel free to use the current nightly build until `3.1.2` has been released. It is stable.

**Comment by dsyalex** [comment]
Thank you! Adding the NuGet.config file from the readme and changing my version to `3.2.0-preview.20421.1` has done the trick.
